version: '3'

services:
  mysql:
    image: mysql:8.0.32
    container_name: mysql
    restart: unless-stopped
    healthcheck:
      # https://stackoverflow.com/a/51641089
      test: mysql --user=root --password=mysql-root-password --execute 'SHOW DATABASES;'
      interval: 15s
      timeout: 5s
      retries: 3
      start_period: 30s
    networks:
      - private
    volumes:
      - mysql-data:/var/lib/mysql
      - mysql-dump:/mnt/dump
      - ./services/database/init:/docker-entrypoint-initdb.d:ro
      - ./services/database/my.cnf:/etc/mysql/conf.d/my.cnf:ro
      - ./services/database/dump.sh:/usr/local/bin/dump:ro
      - ./services/database/restore.sh:/usr/local/bin/restore:ro
    environment:
      MYSQL_ROOT_PASSWORD: mysql-root-password
    cap_add:
      - SYS_NICE # https://stackoverflow.com/questions/55559386/how-to-fix-mbind-operation-not-permitted-in-mysql-error-log
    labels:
      com.datadoghq.ad.check_names: '["mysql"]'
      com.datadoghq.ad.init_configs: '[{}]'
      com.datadoghq.ad.instances: >
        [{
          "server": "mysql",
          "user": "datadog",
          "pass": "datadog-password",
          "options": {
            "extra_status_metrics": true,
            "extra_innodb_metrics": true,
            "extra_performance_metrics": true,
            "schema_size_metrics": true
          }
        }]
      com.datadoghq.ad.logs: '[{"source": "mysql", "service": "mysql"}]'

networks:
  private: {}

volumes:
  mysql-dump: {}
  mysql-data: {}
